Implement support for injection of maps

graemerocher opened the pull request:
This PR adds support for injecting of maps. It attempts to be backwards compatible by first looking up a bean that implements map and then falling back to map resolution. 

An error occurs if multiple beans exist that do not define a qualifier.

graemerocher committed:
Implement support for injection maps

graemerocher committed:
docs

graemerocher committed:
fix sonar issues

dstepanov commented:
So if you’re going to inject a map of something that is not named it will fail? What if only a few of impl are named?

graemerocher commented:
It works for annotation based qualifiers as well. In that case the name is the qualifier name

graemerocher commented:
If you have multiple unamed beans then they would be treated as multiple primaries in which case it would error. This is similar to a `NonUniqueBeanException`. Actually maybe we should throw that instead of `DependencyInjectionException`

graemerocher committed:
include NonUniqueBeanException in trace

dstepanov commented:
Would it behave correctly when one of the beans is primary?

graemerocher commented:
yes in that case the key will be `Primary`

dstepanov commented:
What would happen if you inject JDBC datasources as String/Datasource map? I think the default is the primary one.

graemerocher committed:
Fix keys for EachProperty

graemerocher commented:
good catch, I added a test for this and implemented some changes https://github.com/micronaut-projects/micronaut-core/pull/8305/commits/c6c7f11d24c316e3f7a21ec641c9352c3235e21a

sdelamo reviewed (changes requested):
In addition to the docs you wrote, add a paragraph in the What's new section telling it is now possible to inject a map whose key is the qualifier's name.

dstepanov commented on inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:
```diff
@@ -934,6 +940,58 @@ protected <T> Stream<T> streamOfType(BeanResolutionContext resolutionContext, Cl
+     * @return A map of beans, never {@code null}.
+     * @since 4.0.0
+     */
+    protected <V> @NonNull Map<String, V> mapOfType(@Nullable BeanResolutionContext resolutionContext, @NonNull Argument<V> beanType, @Nullable Qualifier<V> qualifier) {
```
So If I call this with `mapOfType(MyInterface.class, Qualifiers.ANY)` and:
```
@Singleton
class A implements MyInterface {
}
@Singleton
class B implements MyInterface {
}
@Singleton
class C implements MyInterface {
}
```
I will get only one item in the map with `primary` key?

dstepanov replied:
Ok I see it should return `occurs for duplicate keys` because the stream collet wouldn't allow a conflict. It's still strange that this would be OK when only one implementation is defined but will fail with multiple. IMHO correct would be to fail even with one implementation that doesn't specify some criteria like not being named. 

`Multiple beans were found missing a qualifier resulting in duplicate keys` What kind of qualifier is it missing? 

It's confusing to me that in the Micronaut codebase there are two understandings of what qualifier means:
1. Bean context uses it as a filter
2. Bean definition uses is as a set of `@Qualifier`s defined

graemerocher committed:
allow maps of beans without qualifiers

graemerocher commented:
@dstepanov I pushed a change that allows maps of beans without qualifiers by using the simple name of the class decaptilized as the bean name by default if no qualifier is present. This appears to be what Spring does as well.

This also matches up with how we qualify since if no qualifier is present we use the simple name of the type. See https://github.com/micronaut-projects/micronaut-core/blob/686b472bc4801887a115c5d06cb015301241875b/inject/src/main/java/io/micronaut/inject/qualifiers/NameQualifier.java#L89

dstepanov commented on inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:
```diff
@@ -986,10 +987,19 @@ protected <T> Stream<T> streamOfType(BeanResolutionContext resolutionContext, Cl
         if (definition instanceof NameResolver resolver) {
-            return resolver.resolveName().orElse(reg.identifier.getName());
+            return resolver.resolveName().orElse(identifier.getName());
+        } else {
```
You might want to avoid redundant `else` here and below.

graemerocher committed:
checkstyle
